04-advanced-topics/network-analysis_rnaseq_01_wgcna.Rmd

```diff
@@ -420,13 +425,267 @@ readr::write_rds(bwnet,
+  #
+  # Returns:
+  # A module's genes in heatmap with the module's eigengenes in a barplot
+
+  # Set up column annotation from metadata
+  col_annot_df <- metadata_df %>%
+    # Only select the treatment and sample ID columns
+    dplyr::select(refinebio_accession_code, refinebio_treatment) %>%
+    # Add on the eigengene expression as its own column
+    dplyr::mutate(
+      module_eigengene = module_eigengenes_df %>%
+        dplyr::pull(module_name)
+    ) %>%
+    # Store sample
+    tibble::column_to_rownames("refinebio_accession_code") %>%
+    # Arrange by treatment
+    dplyr::arrange(refinebio_treatment)
+
+  # Create the ComplexHeatmap column annotation object
+  col_annot <- ComplexHeatmap::HeatmapAnnotation(
+    # Supply treatment labels
+    refinebio_treatment = col_annot_df$refinebio_treatment,
+    # Add annotation barplot
+    module_eigengene = ComplexHeatmap::anno_barplot(col_annot_df$module_eigengene),
+    # Pick colors for each experimental group in refinebio_treatment
+    col = list(refinebio_treatment = c("post-adt" = "#f1a340", "pre-adt" = "#998ec3"))
+  )
+
+  # Get a vector of the Ensembl gene ides that correspond to this module
+  module_genes <- gene_module_key_df %>%
+    dplyr::filter(module == module_name) %>%
+    dplyr::pull(gene)
+
+  # Set up the gene expression data frame
+  mod_mat <- expression_mat %>%
+    t() %>%
+    as.data.frame() %>%
+    # Only keep genes from this module
+    dplyr::filter(rownames(.) %in% module_genes) %>%
+    # Order the samples to match col_annot_df
+    dplyr::select(rownames(col_annot_df)) %>%
+    # Data needs to be a matrix
+    as.matrix()
+
+  if (standardize_genes) {
+    # Genes are on very different scales; we will standardize them
+    mod_mat <- t(apply(mod_mat, 1, scale))
+
+    # Create a color function based on standardized scale
+    color_func <- circlize::colorRamp2(
+      c(-1, 0, 1),
+      c("#67a9cf", "#f7f7f7", "#ef8a62")
+    )
+  } else {
+    # Create a color function based on standardized scale
+    color_func <- circlize::colorRamp2(
+      c(
+        quantile(mod_mat, 0.05),
+        median(mod_mat),
+        quantile(mod_mat, 0.95)
+      ),
+      c("#67a9cf", "#f7f7f7", "#ef8a62")
+    )
+  }
+
+  # Plot on a heatmap
+  heatmap <- ComplexHeatmap::Heatmap(mod_mat,
+    name = module_name,
+    # Supply color function
+    col = color_func,
+    # Supply column annotation
+    bottom_annotation = col_annot,
+    # We don't want to cluster samples
+    cluster_columns = FALSE,
+    # We don't need to show sample or gene labels
+    show_row_names = FALSE,
+    show_column_names = FALSE
+  )
+
+  # Return heatmap
+  return(heatmap)
+}
+```
+
+## Make module heatmaps
+
+Let's try out the custom heatmap function with module 52 (our most differentially expressed module).
+
+```{r}
+mod_52_heatmap <- make_module_heatmap(module_name = "ME52")
+
+# Print out the plot
+mod_52_heatmap
```

Comment:
I don't love that the legends are side by side instead of one above the other, but the fix seems like it adds way too much complexity, so I would rather live with this.